text/0000-label-break-value.md

```diff
@@ -0,0 +1,119 @@
+- Feature Name: label_break_value
+- Start Date: 2017-06-26
+- RFC PR: [#2046](https://github.com/rust-lang/rfcs/pull/2046)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Allow a `break` of labelled blocks with no loop, which can carry a value.
+
+# Motivation
+[motivation]: #motivation
+
+In its simplest form, this allows you to terminate a block early, the same way that `return` allows you to terminate a function early.
+
+```
+'block: {
+    do_thing();
+    if condition_not_met() {
+        break 'block;
+    }
+    do_next_thing();
+    if condition_not_met() {
+        break 'block;
+    }
+    do_last_thing();
+}
+```
+In the same manner as `return` and the labelled loop breaks in [RFC 1624](https://github.com/rust-lang/rfcs/blob/master/text/1624-loop-break-value.md), this `break` can carry a value:
+```
+let result = 'block: {
+    if foo() { break 'block 1; }
+    if bar() { break 'block 2; }
+    3
+};
+```
+RFC 1624 opted not to allow options to be returned from `for` or `while` loops, since no good option could be found for the syntax, and it was hard to do it in a natural way. This proposal gives us a natural way to handle such loops with no changes to their syntax:
+```
+let result = 'block: {
+    for &v in container.iter() {
+        if v > 0 { break 'block v; }
+    }
+    0
+};
+```
+This extension handles searches more complex than loops in the same way:
+```
+let result = 'block: {
+    for &v in first_container.iter() {
+        if v > 0 { break 'block v; }
+    }
+    for &v in second_container.iter() {
+        if v < 0 { break 'block v; }
+    }
+    0
+};
+```
+Implementing this without a labelled break is much less clear:
+```
+let mut result = None;
+for &v in first_container.iter() {
+    if v > 0 {
+        result = Some(v);
+        break;
+    }
+}
+if result.is_none() {
+    for &v in second_container.iter() {
+        if v < 0 {
+            result = Some(v);
+            break;
+        }
+    }
+}
+let result = result.unwrap_or(0);
+```
+
+# Detailed design
+[design]: #detailed-design
+```
+'BLOCK_LABEL: { EXPR }
+```
+would simply be syntactic sugar for
+```
+'BLOCK_LABEL: loop { break { EXPR } }
+```
+except that unlabelled `break`s or `continue`s which would bind to the implicit `loop` are forbidden inside the *EXPR*.
```

Comment:
At first I was worried about forbidding unlabelled `break`s or `continue`s, since it would mean that adding a fresh block label to existing code could cause that code to stop compiling, like so:

Old code that compiles:
```rust
loop {
    {  
        let raii_thing = Destro::new();
        if gi_joe() { break; }
    }
    println!("no more Destro here!");
}
```

New code that would not compile:

```rust
loop {
    'fresh_unused_label_for_expository_text_describing_raii: {
        let raii_thing = Destro::new();
        if gi_joe() { break; }
        //~^ ERROR: unlabelled break within labelled block with no intervening loop syntax.
    }
    println!("no more Destro here!");
}
```

However, after a bit of thought (and reading over the alternatives section that explicitly explains *why* this RFC is choosing this conservative approach), I decided that forbidding unlabelled `break`s/`continue`s is acceptable *as long as* the `rustc` error messages suggest to the user that they explicitly label their break target in such a case.

E.g., I want the compiler, when given the above code that fails to compile, to suggest to the programmer that they write something analogous to the following:

```rust
'escape_hatch: loop {
    'fresh_unused_label_for_expository_text_describing_raii: {
        let raii_thing = Destro::new();
        if gi_joe() { break 'escape_hatch; }
    }
    println!("no more Destro here!");
}
```

I suspect a high quality error message in such cases will required for the success of this feature, so I'd go so far as to say that the RFC itself should mention it.

Reply:
Thanks! I've had a go at adding some language, let me know what you think. Note that I'm offline from Sunday 27th to Tuesday 5th.